utils/char16trie/src/char16trie.rs

```diff
@@ -0,0 +1,396 @@
+///  - Branch node: Branches to other nodes according to the current input unit.
+///    The node unit is the length of the branch (number of units to select from)
+///    minus 1. It is followed by a sub-node:
+///    - If the length is at most MAX_BRANCH_LINEAR_SUB_NODE_LENGTH, then
+///      there are length-1 (key, value) pairs and then one more comparison unit.
+///      If one of the key units matches, then the value is either a final value for
+///      the string so far, or a "jump" delta to the next node.
+///      If the last unit matches, then matching continues with the next node.
+///      (Values have the same encoding as final-value nodes.)
+///    - If the length is greater than MAX_BRANCH_LINEAR_SUB_NODE_LENGTH, then
+///      there is one unit and one "jump" delta.
+///      If the input unit is less than the sub-node unit, then "jump" by delta to
+///      the next sub-node which will have a length of length/2.
+///      (The delta has its own compact encoding.)
+///      Otherwise, skip the "jump" delta to the next sub-node
+///      which will have a length of length-length/2.
+///
+/// For more information:
+/// - [ICU4C UCharsTrie](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/classicu_1_1UCharsTrie.html)
+/// - [ICU4J CharsTrie](https://unicode-org.github.io/icu-docs/apidoc/released/icu4j/com/ibm/icu/util/CharsTrie.html) API.
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[derive(Clone)]
+pub struct Char16Trie<'data> {
+    /// An array of u16 containing the trie data.
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub data: ZeroVec<'data, u16>,
+}
+
+/// This struct represents an iterator over a Char16Trie.
+pub struct Char16TrieIterator<'a> {
+    /// A reference to the Char16Trie data to iterate over.
+    trie: &'a [<u16 as zerovec::ule::AsULE>::ULE],
+    /// Index of next trie unit to read, or None if there are no more matches.
+    pos: Option<usize>,
+    /// Remaining length of a linear-match node, minus 1, or None if not in
+    /// such a node.
+    remaining_match_length: Option<usize>,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum TrieResult {
+    /// The input unit(s) did not continue a matching string.
+    /// Once next() return TrieResult::NoMatch, all further calls to next()
+    /// will also return TrieResult::NoMatch.
+    NoMatch,
+    /// The input unit(s) matched a string but there is no value for the string
+    /// so far.  (It is a prefix of a longer string.)
+    NoValue,
+    /// The input unit(s) continued a matching string and there is a value for
+    /// the string so far. No further input byte/unit can continue a matching
+    /// string.
+    FinalValue(i32),
+    /// The input unit(s) continued a matching string and there is a value for
+    /// the string so far.  Another input byte/unit can continue a matching
+    /// string.
+    Intermediate(i32),
+}
+
+impl<'a> Char16TrieIterator<'a> {
```

Comment:
There is very little of the public API here, compared to the ICU version. I understand that the ICU CharsTrie.Iterator (which iterates over the (string, value) pairs reachable from the current state) is out of scope here, but what about the rest?

https://unicode-org.github.io/icu-docs/apidoc/released/icu4j/com/ibm/icu/util/CharsTrie.html

At a minimum, I would expect to see nextForCodePoint(int cp):
https://github.com/unicode-org/icu/blob/main/icu4j/main/classes/core/src/com/ibm/icu/util/CharsTrie.java#L250-L263

I think Henri will need that for collation.

And since you intentionally omit API that is well-defined and exists in ICU, I think that somewhere in this file there should be a note pointing to ICU for the complete set.

Reply:
This code started with what @makotokato implemented to support segmentation, and added what @hsivonen asked for to support the collator. I wasn't trying for a complete port, just trying to address our immediate needs.

Since we're not at 1.0 yet, we can expand the API later as needed. But maybe we should just do it now. I'd be interested in other's thoughts on this as well.

Reply:
The most important method is `nextForCodePoint()`.  In fact, the method labeled `next()` is dangerous, because it looks like you can pass a code point in there, but as soon as you hit a supplementary code point, you will get unexpected results.

I suggest the following methods:

1. `next(char cp)`
3. `next_u16(u16 code_unit)`
2. `next_u32(u32 cp)`

Reply:
The data structure is defined as mapping from a sequence of 16-bit units to an int32_t. It should support that, even when the 16-bit units have nothing to do with UTF-16 strings, and so next(unit) should be the primary entry point.

"nextCodePoint()" is named suggestively, but there is no enforcement that the inputs must be Unicode strings; it's for a gentleman's agreement between data builder and user to store and provide surrogate pairs.

I found it very handy for next() to take an int32_t with values outside 0..ffff simply not matching: Neither the caller nor the implementation needs to do a range check, error values simply yield NO_MATCH. Easy to use, and efficient.

Remember that this is a low-level data structure.

Reply:
Idea:
- TrieResult is just a four-value enum like in C++ & Java; no built-in value
- no public get_value()
- next(unit) returns TrieResult, no way to get the value
- next_with_value(unit) returns a (TrieResult, value) pair
- next_code_point(cp) calls next(lead surrogate) + next(trail surrogate)
- next_code_point_with_value(cp) calls next(lead surrogate) + next_with_value(trail surrogate)
- the not-with-value functions *could* eagerly skip the value units, if the node has any

Reply:
This also seems like a reasonable suggestion to me. I actually don't have that strong of an opinion on the API surface, but I would like for us to come to a decision on this before I write any more code :)

@sffc what do you think?

Reply:
I don't feel strongly about when TrieResult and the value are or are not returned.  I just want to avoid the footgun of the default `next()` method taking a `u32` of which only 16 bits are ever used.

Since it seems like we can't agree on that, can we agree on suffixing all of the `next` methods?

- `next_u16(u16)`
- `next_code_point(u32)`
- `next_char(char)`

Reply:
Discussion 2021-12-09:

- Rust may perform optimizations over what C++ had done
- C++ APIs are less ergonomic / have footguns
- Conclusion: Start with Rusty APIs, and file a follow-up issue to investigate codegen and experiment with Markus's suggestions

Reply:
I am ok with suffixes for the input type. How about the following then:
- TrieResult is just a four-value enum like in C++ & Java; no built-in value
- no public get_value()
- next_for_u16(u16 unit) returns TrieResult, no way to get the value
- next_for_char(char c) calls next(lead surrogate) + next(trail surrogate) if supplementary, and returns TrieResult, no way to get the value
- next_with_value_for_u16(unit) returns a (TrieResult, value) pair
- next_with_value_for_char(char c) ...
- the not-with-value functions *could* eagerly skip the value units, if the node has any

From much work in C/C++/Java, where you frequently end up with "int" anyway, I would also want ..._for_i32(i32_unit) versions of these functions; same as the ..._for_u16(u16 unit) versions except if the unit is outside 0..ffff then no match. Note that in the trie implementation this happens automatically -- no range checks needed on the unit. I don't know whether the same kind of coding leads Rust devs to end up with i32 or u32 types, so we could omit these versions for now and see if call sites end up with inconvenient and unnecessary range checks.

Reply:
My plan is to land this in experimental for now, since it has been difficult for me to find the time to work on this and this will unblock other people from using this code. I've filed https://github.com/unicode-org/icu4x/issues/1439 to track the performance related points that Markus brought up during the review.